DL3-35 secure middleware courses endpoint

mgwozdz-unicon opened the pull request:
## Description
Secured the middleware's backend endpoint for retrieving the Harmony course catalog using the LTI id_token.

### Motivation and Context
Only users who have a valid LTI registration who are accessing the Deep Linking menu through a valid LTI flow should be permitted to access the course catalog via the middleware.

### Fixes
[DL3-35](https://lumenlearning.atlassian.net/browse/DL3-35)

## How Has This Been Tested?
0. Ensure as a prerequisite that the tool was registered and deployed in D2L and that a Deep Linking Quicklink Link has been added to the Deployment and that there is an org unit appropriately given permission to access the tool.
1. Go to the course (org unit) that has permission to access the tool in D2L.
2. Click on Content in the top navigation.
3. Click "Add a module...", write a name for it, and hit enter to create the module.
4. Click the dropdown for Existing Activities
5. Click on the Deep Linking Quicklink (e.g. I named mine "Lumen Content Selector").
6. Ensure that the course catalog displays with the expected data being sent from the Harmony /course_catalog endpoint and that the pagination is in place if there are more than 10 courses.
Note: If all of the courses in the catalog are using the brain image and were released on 1/18/2022, then this is an indication that the call to Harmony failed and test data has been returned.

## Screenshots
N/A
---
## Database changes
N/A

## ⚠️ Deployment instructions ⚠️
N/A

## New ENV variables
N/A

---

## Checklist
- [x] I have reviewed the AC for this feature and my changes meet all requirements
- [x] (For UI changes) I have considered the accessibility of this feature (see https://www.a11yproject.com/checklist/ for a high-level checklist)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have requested a review from at least one other dev

leannethng reviewed (approved):
Thank you Mary!

smith750 commented on src/main/frontend/src/app/appSlice.js:
```diff
@@ -20,13 +32,25 @@ export const appSlice = createSlice({
       state.metadata = action.payload;
     },
     changeSearchInput: (state, action) => {
+      debugger;
```
I'm assuming this line will be removed before the PR is out of draft?

smith750 commented on src/main/java/net/unicon/lti/controller/harmony/HarmonyRestController.java:
```diff
@@ -45,26 +36,24 @@ public class HarmonyRestController {
+            // We convert the JSON response to a Java object, and send the JSON value again to the frontend.
+            return ResponseEntity.ok(harmonyService.fetchHarmonyCourses());
+        } catch (Exception e) {
+            e.printStackTrace();
```
Should this go through the logger?

smith750 reviewed (approved):
I made some nit-picky comments but this looks fantastic.  Thank you!

HotchkissLuLe reviewed (approved):
This looks great, thanks Mary!

mpellicer commented on src/main/frontend/src/app/appSlice.js:
```diff
@@ -50,7 +74,10 @@ export const selectLoading = (state) => state.loading;
-  fetch('/harmony/courses')
+  fetch('/harmony/courses', {
+    method: 'GET',
+    headers: {'lti-id-token': initialState.id_token}
```
We can use the state, state.id_token

mpellicer commented on src/main/frontend/src/app/appSlice.js:
```diff
@@ -1,12 +1,24 @@
 export const appSlice = createSlice({
   // Name of the slice
   name: 'appSlice',
+  // Initial state
```
The initial state is defined in index.js and it's already inserting the id_token

mpellicer commented on src/main/frontend/src/app/appSlice.js:
```diff
@@ -1,12 +1,24 @@
-// Static data
-import { STATIC_COURSE_ARRAY } from './staticData';
+// Parse the LtiLaunchData JSON object from the root attribute.
+const ltiLaunchData = JSON.parse(document.getElementById('root').getAttribute('lti-launch-data'));
```
This is done in the index.js, we can use the token.

mpellicer commented on src/main/frontend/src/app/appSlice.js:
```diff
@@ -1,12 +1,24 @@
 import { createSlice } from '@reduxjs/toolkit';
+import { STATIC_COURSE_ARRAY } from './staticData.js';
```
This is not needed, in DL3-34 it simulates pagination and is updated.

mgwozdz-unicon commented:
I believe I've addressed all of the comments and set this as ready for review. Let me know if there are any additional changes I should make. Thank you!

smith750 commented:
Thank you for the updates Mary!
